src/test/java/com/cst438/controller/ControllerUnitTests.java

```diff
@@ -0,0 +1,151 @@
+    @Autowired
+    MockMvc mvc;
+
+    @Autowired
+    SectionRepository sectionRepository;
+
+    @Autowired
+    AssignmentRepository assignmentRepository;
+
+    @Autowired
+    GradeRepository gradeRepository;
+
+    @Autowired
+    EnrollmentController enrollmentController;
+
+    // Unit Test #4 - Instructor grades an assignment, enters scores for all enrolled students, and uploads the scores
+    @Test
+    public void gradeAssignment() throws Exception {
+        MockHttpServletResponse response;
+
+        response = mvc.perform(
+                        MockMvcRequestBuilders
+                                .get("/assignments/1/grades")
+                                .accept(MediaType.APPLICATION_JSON)
+                                .contentType(MediaType.APPLICATION_JSON))
+                        .andReturn()
+                        .getResponse();
+
+        assertEquals(200, response.getStatus());
+
+        GradeDTO[] result = fromJsonString(response.getContentAsString(), GradeDTO[].class);
+        int numDTOs = result.length;
+
+        // List of grades grabbed from database shouldn't be empty.
+        assertNotEquals(0, numDTOs);
+
+        /*
+         * Pulling old scores, altering them, and saving a new list of GradeDTOs.
+         */
+        Integer[] oldScores = new Integer[numDTOs];
+        Integer[] newScores = new Integer[numDTOs];
+        GradeDTO[] newDTOs = new GradeDTO[numDTOs];
+        for (int i = 0; i < numDTOs; i++) {
+            GradeDTO target = result[0];
+            oldScores[i] = target.score();
+
+            int newScore = oldScores[i] - 10;
+            if (newScore < 0)
+                newScore += 100;
+
+            newScores[i] = newScore;
+            newDTOs[i] = new GradeDTO(target.gradeId(), target.studentName(), target.studentEmail(), target.assignmentTitle(), target.courseId(), target.sectionId(), newScores[i]);
+        }
+
+        // Confirming that all scores were altered successfully.
+        for (int i = 0; i < numDTOs; i++) {
+            assertNotEquals(newScores[i], oldScores[i]);
+            assertNotEquals(oldScores[i], newDTOs[i].score());
+            assertEquals(newScores[i], newDTOs[i].score());
+        }
+
+        // Saving all altered GradeDTOs.
+        response = mvc.perform(
+                        MockMvcRequestBuilders
+                                .put("/grades")
+                                .accept(MediaType.APPLICATION_JSON)
+                                .contentType(MediaType.APPLICATION_JSON)
+                                .content(asJsonString(newDTOs)))
+                        .andReturn()
+                        .getResponse();
+        
+        assertEquals(200, response.getStatus());
+
+        // Confirming that all DTOs were saved successfully.
+        for (int i = 0; i < numDTOs; i++) {
+            Grade g = gradeRepository.findById(newDTOs[i].gradeId()).orElse(null);
+            assertEquals(g.getScore(), newScores[i]);
```

Comment:
Possible `NullPointerException`; since this is a test that is fine, but maybe a proper assert should be coded.